lib/src/custom_rule_generator.dart

```diff
@@ -6,43 +6,190 @@
+    "packages/serialization/src/generated_serialization_rules.dart";
+
+/// This string can be found in the generated file. It indicates the location
+/// where the imports required for the generated code must be appended.
+const String GENERATED_IMPORTS_MARKER = "//{{GENERATED_IMPORTS_MARKER}}";
+
+/// This string can be found in the generated file. It indicates the location
+/// where the generated rules must be appended.
+const String GENERATED_RULES_MARKER = "//{{GENERATED_RULES_MARKER}}";
+
+/// This string can be found in the generated file. It indicates the location
+/// where the [Map] of generated rules must be appended.
+const String GENERATED_RULES_MAP_MARKER =  "//{{GENERATED_RULES_MAP_MARKER}}";
+
+/// This string can be found in the generated rules and imports. It is used as
+/// the name of an `as` in the import statement to avoid import conflicts.
+const String AS_PLACEHOLDER = "{{AS_PLACEHOLDER}}";
 
 /// Given the [contents] of a Dart source library, generate
-/// [CustomRule]s hard-coded for those classes and return a string
-/// representing the source of those new rules.
+/// [CustomRule]s hard-coded and other information for those classes and return.
 ///
 /// If [listFormat] is true, then the output of the rule will be a list
 /// of field values. If it is false, then the output is a map keyed by
-/// field name. The [libraryName] is used as the name of the
-/// generated library. It will import [originalImport] and expects that to
-/// contain the classes in [contents]. See the comment for transformer.dart.
-String generateCustomRulesFor(String contents,
-    {String libraryName, String originalImport, bool listFormat : true}) {
+/// field name. See the comment for transformer.dart.
+AssetSerializationAnalysisResults analyzeAsset(String contents,
+    {bool listFormat: true, processAnnotatedClasses: false,
+    processAllClasses: false}) {
 
   var lib = parseCompilationUnit(contents);
-  var classes = lib.declarations.where((x) => x is ClassDeclaration);
-  var rules =
-      classes.map((each) => new CustomRuleGenerator(each, listFormat)).toList();
-  var mapOfRuleNamesToRules = rules
-      .map((x) => "    '${x._declaration.name}' : new ${x.ruleName}()")
-      .join(",\n");
-  var ruleDeclarations = rules.map((x) => x.rule).join("\n\n");
 
-  return '''
-// Generated serialization rules. *** DO NOT EDIT ***
-// See transformer.dart in package serialization.
-library ${libraryName}_serialization_rules;
+  // Check if the file is a part of another or is the main library file and find
+  // the library name.
+  String libraryName;
+  PartOfDirective partOf = lib.directives.firstWhere(
+          (x) => x is PartOfDirective, orElse: () => null);
+  if(partOf != null) {
+    libraryName = "${partOf.libraryName}";
+  } else {
+    LibraryDirective library = lib.directives.firstWhere(
+            (x) => x is LibraryDirective, orElse: () => null);
+    libraryName = library == null ? "__default" : "${library.name}";
```

Comment:
Why the interpolation. Just library.name? Or even library.name.toString() if it's some analyzer structure.